references/cuegen/generators/provider/provider.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright 2023 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	goast "go/ast"
+	"io"
+	"strings"
+
+	cueast "cuelang.org/go/cue/ast"
+	cuetoken "cuelang.org/go/cue/token"
+	"golang.org/x/tools/go/packages"
+
+	"github.com/oam-dev/kubevela/references/cuegen"
+)
+
+const (
+	typeProviderFnMap          = "map[string]github.com/kubevela/pkg/cue/cuex/runtime.ProviderFn"
+	typeProvidersParamsPrefix  = "github.com/kubevela/pkg/cue/cuex/providers.Params"
+	typeProvidersReturnsPrefix = "github.com/kubevela/pkg/cue/cuex/providers.Returns"
+)
+
+const (
+	doKey       = "do"
+	providerKey = "provider"
+)
+
+type provider struct {
+	name    string
+	params  string
+	returns string
+	do      string
+}
+
+// Options is options of generation
+type Options struct {
+	File     string    // Go file path
+	Writer   io.Writer // target writer
+	AnyTypes []string  // option cuegen.WithAnyTypes
+	Nullable bool      // option cuegen.WithNullable
+}
+
+// Generate generates cue provider from Go struct
+func Generate(opts Options) (rerr error) {
+	g, err := cuegen.NewGenerator(opts.File)
+	if err != nil {
+		return err
+	}
+
+	// make options
+	genOpts := make([]cuegen.Option, 0)
```

Comment:
Maybe it's better to have a unify way to convert `Option` to `cuegen.Option`.

Reply:
I can't get the point. provider-gen only exposes some options, so it can prevent options such as type filter from being modified by the caller.